roomserver/internal/query/query_room_hierarchy.go

```diff
@@ -0,0 +1,597 @@
+package query
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"strings"
+
+	fs "github.com/matrix-org/dendrite/federationapi/api"
+	"github.com/matrix-org/dendrite/internal/caching"
+	roomserver "github.com/matrix-org/dendrite/roomserver/api"
+	"github.com/matrix-org/dendrite/roomserver/types"
+	userapi "github.com/matrix-org/dendrite/userapi/api"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/gomatrixserverlib/fclient"
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/matrix-org/util"
+	"github.com/tidwall/gjson"
+)
+
+// Query the hierarchy of a room (A.K.A 'space summary')
+//
+// This function returns a iterator-like struct over the hierarchy of a room.
+func (r *Queryer) QueryRoomHierarchy(ctx context.Context, caller types.DeviceOrServerName, roomID spec.RoomID, suggestedOnly bool, maxDepth int) roomserver.RoomHierarchyWalker {
+	walker := RoomHierarchyWalker{
+		rootRoomID:         roomID.String(),
+		caller:             caller,
+		thisServer:         r.Cfg.Global.ServerName,
+		rsAPI:              r,
+		fsAPI:              r.FSAPI,
+		ctx:                ctx,
+		roomHierarchyCache: r.Cache,
+		suggestedOnly:      suggestedOnly,
+		maxDepth:           maxDepth,
+		unvisited: []roomVisit{{
+			roomID:       roomID.String(),
+			parentRoomID: "",
+			depth:        0,
+		}},
+	}
+
+	return &walker
+}
+
+type stringSet map[string]struct{}
+
+func (s stringSet) contains(val string) bool {
+	_, ok := s[val]
+	return ok
+}
+
+func (s stringSet) add(val string) {
+	s[val] = struct{}{}
+}
+
+type RoomHierarchyWalker struct {
+	rootRoomID         string // TODO change to spec.RoomID
+	caller             types.DeviceOrServerName
+	thisServer         spec.ServerName
+	rsAPI              *Queryer
+	fsAPI              fs.RoomserverFederationAPI
+	ctx                context.Context
+	roomHierarchyCache caching.RoomHierarchyCache
+	suggestedOnly      bool
+	maxDepth           int
+
+	processed stringSet
+	unvisited []roomVisit
+
+	done bool
+}
+
+const (
+	ConstCreateEventContentKey        = "type"
+	ConstCreateEventContentValueSpace = "m.space"
+	ConstSpaceChildEventType          = "m.space.child"
+	ConstSpaceParentEventType         = "m.space.parent"
+)
```

Comment:
These should be moved to GMSL: https://github.com/matrix-org/gomatrixserverlib/blob/main/spec/eventtypes.go

Reply:
Opened PR here https://github.com/matrix-org/gomatrixserverlib/pull/402